### PR TITLE
Clean up image upload and retrieval logic inside of a program as an admin

### PR DESCRIPTION
**Pull request by innopoop**
Addresses #88 

Routes affected:
/program/[slug]/
/program/[slug]/settings

Changes:
1. Settings page icon upload & saving
2. Homepage image proper resizing
3. Homepage image proper upload
4. Homepage image proper retrieval and restoration of dimensions
5. Styling changes

### CODE REVIEW

**Comment by nova-admin** [comment]
#### 📝Changed routes:


| [**/program/\[slug\]**](https://risc-lmhym6blq-novaforgood.vercel.app/program/[slug]) | [**/program/\[slug\]/settings**](https://risc-lmhym6blq-novaforgood.vercel.app/program/[slug]/settings) |
|:-:|:-:|
|<a href="https://risc-lmhym6blq-novaforgood.vercel.app/program/[slug]">Screenshot of /program/[slug]</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/xOa5h_jvQwTtKJJwbx7JFXFX8bN-1m0OhiAM_XVTJeg-w_NKqOVV27X5R53QAnim1Pw36PxzM3126yQhpfp9Fw.png">(view full size)</a> |<a href="https://risc-lmhym6blq-novaforgood.vercel.app/program/[slug]/settings">Screenshot of /program/[slug]/settings</a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/ocoJ5qvcrB8NUfBg6Gzr8xsfU_4z5tRS14Jf2MW0m6T6PycG7QAb_8CqYgmGz_m5HsVfkfsp-1TU2KQoeZo19g.png">(view full size)</a> |







Commit a1ba467944ef05007c7ccce26c39b73cea4986d9 (https://risc-lmhym6blq-novaforgood.vercel.app).

**Comment by legitmaxwu** [comment]
Good stuff! Icon upload works great, and I'll leave code comments later.

Comments on general functionality:

- Cannot edit homepage text (I type and nothing shows up)
- Image sizing should be more sensitive (I have to drag my cursor a lot just to change the image size a little)
- No point in having the drag handles on the middles of edges because you can't squash the image
- Adding an image does not count as editing the homepage (Publish button still disabled)

**Comment by legitmaxwu** [comment]
image

Also, could you look into the dropdown image not changing on program icon change? (Should just be about adding a refetch or something)

**Comment by innopoop** [comment]
> 
> 
> Also, could you look into the dropdown image not changing on program icon change? (Should just be about adding a refetch or something)

I looked into why the tab wasn't showing the correct program and it was weird because even though I refetched inside of the component, it was getting the wrong data altogether. I'm not sure if it has to do with the fact that Choose Tab Layout is only rendered once as a child of the page or something else.

**Comment by legitmaxwu** [comment]
> > 
> > Also, could you look into the dropdown image not changing on program icon change? (Should just be about adding a refetch or something)
> 
> I looked into why the tab wasn't showing the correct program and it was weird because even though I refetched inside of the component, it was getting the wrong data altogether. I'm not sure if it has to do with the fact that Choose Tab Layout is only rendered once as a child of the page or something else.

Did you refetch getMyUser or getProgramBySlug?

**Review by legitmaxwu — CHANGES_REQUESTED** [review]
Editing works now! More functional comments:

- If I add an image, hitting publish does not save anything properly
- Adding an image still doesn't count as an "edit"

**Comment by innopoop** [comment]
> > > 
> > > Also, could you look into the dropdown image not changing on program icon change? (Should just be about adding a refetch or something)
> > 
> > 
> > I looked into why the tab wasn't showing the correct program and it was weird because even though I refetched inside of the component, it was getting the wrong data altogether. I'm not sure if it has to do with the fact that Choose Tab Layout is only rendered once as a child of the page or something else.
> 
> Did you refetch getMyUser or getProgramBySlug?

I refetched both just to check, but I think you only need to refetch program correct? Perhaps refetching it inside of the ProgramDropdown component isn't the best idea. I will need to think about how to update the component a little more.

File: src/pages/program/[slug]/settings.tsx [hunk 1]
```diff
@@ -171,19 +181,37 @@ const SettingsPage: Page = () => {
...
                   programId,
                   data: {
+                    iconUrl: newIconUrl,
                     name: mentorshipName,
                     description: mentorshipDescription,
                   },
                 },
               }).then(() => {
                 refetchCurrentProgram();
+                refetchGetMyUserQuery();
```
Comment (legitmaxwu) [inline src/pages/program/[slug]/settings.tsx:209]: Delete this line

File: src/pages/program/[slug]/settings.tsx [hunk 2]
```diff
@@ -60,6 +62,7 @@ const SettingsPage: Page = () => {
   };
 
   //TODO: Image Upload and Resize => URL save
+  const [uploadImageAndResizeMutation] = useUploadImageAndResizeMutation();
   const [updateProgram] = useUpdateProgramMutation();
```
Comment (legitmaxwu) [inline src/pages/program/[slug]/settings.tsx:66]:
```
useUpdateProgramMutation({ refetchQueries: [ refetchGetMyUserQuery() ] });
```

**Review by legitmaxwu — CHANGES_REQUESTED** [review]
I found the reason why it's not refetching--the imported refetch function from graphql.ts is not meant for direct calling, it's meant to be passed into the refetchQueries parameter in the useMutation hook.

If you would like to call refetch directly, you'd need to unpack the function from the useQuery hook. i.e. `const { data, refetch } = useQuery()`

See:
- https://www.apollographql.com/docs/react/data/queries/#refetching
- https://www.apollographql.com/docs/react/caching/advanced-topics/
- https://www.graphql-code-generator.com/docs/generated-config/typescript-react-apollo#withrefetchfn

File: src/pages/program/[slug]/settings.tsx (hunk 2 above)
Comment (innopoop) [reply src/pages/program/[slug]/settings.tsx:66]: brain is yuge

**Review by legitmaxwu — APPROVED** [review]

**Pull request merged**